staging/src/k8s.io/kubectl/pkg/cmd/apply/applyset_pruner.go

```diff
@@ -16,15 +16,114 @@ limitations under the License.
 
 package apply
 
-import "fmt"
+import (
+	"context"
+	"fmt"
+	"io"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/cli-runtime/pkg/printers"
+	"k8s.io/client-go/dynamic"
+	cmdutil "k8s.io/kubectl/pkg/cmd/util"
+)
 
 type applySetPruner struct {
+	dynamicClient dynamic.Interface
+
+	visitedUids sets.String
+
+	cascadingStrategy metav1.DeletionPropagation
+	dryRunStrategy    cmdutil.DryRunStrategy
+	gracePeriod       int
+
+	toPrinter func(string) (printers.ResourcePrinter, error)
+
+	out io.Writer
+}
+
+func newApplySetPruner(o *ApplyOptions) *applySetPruner {
+	return &applySetPruner{
+		dynamicClient: o.DynamicClient,
+
+		visitedUids: o.VisitedUids,
+
+		cascadingStrategy: o.DeleteOptions.CascadingStrategy,
+		dryRunStrategy:    o.DryRunStrategy,
+		gracePeriod:       o.DeleteOptions.GracePeriod,
+
+		toPrinter: o.ToPrinter,
+
+		out: o.Out,
+	}
 }
 
-func newApplySetPruner(_ *ApplyOptions) *applySetPruner {
-	return &applySetPruner{}
+func (p *applySetPruner) pruneAll(ctx context.Context, applyset *ApplySet) error {
+	applysetLabelSelector := applyset.LabelSelectorForMembers()
+
+	opt := metav1.ListOptions{
+		LabelSelector: applysetLabelSelector,
+	}
+
+	// TODO: Split into discovery and deletion, run discovery in parallel (and maybe in consistent order or in parallel?)
+	for _, restMapping := range applyset.AllPrunableResources() {
+		switch restMapping.Scope.Name() {
+		case meta.RESTScopeNameNamespace:
+			for _, namespace := range applyset.AllPrunableNamespaces() {
+				if err := p.prune(ctx, namespace, restMapping, opt); err != nil {
+					return fmt.Errorf("error pruning namespaced object %v: %w", restMapping.GroupVersionKind, err)
```

Comment:
Can we get test coverage for what this error looks like? We're talking about an individual object but the GVR injection at this point won't name it, so we're assuming the error we got from `p.prune` does (and that it doesn't already name the GVR)

Reply:
Added a test case and cleaned the message up a bit...

Here's what it looks like right now:

```
namespace/foo unchanged
error: pruning Namespace objects: deleting bar: an error on the server ("") has prevented the request from succeeding
```

I don't love that wording (despite my tweaking it already), I think maybe we should combined the GVK with the NN in the message.  WDYT?